src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+      int sampleSize) {
+    int[] distribution = new int[tupleCounts.size()];
+    int totalTupleCount = 0;
+    for (int val : tupleCounts) {
+      totalTupleCount += val;
+    }
+
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (int j = 0; j < tupleCounts.size(); j++) {
+        if (sampleTupleIdx < tupleCounts.get(j) + tupleOffset) {
+          distribution[j] += 1;
+          break;
+        }
+        tupleOffset += tupleCounts.get(j);
+      }
+    }
+    return distribution;
+  }
+
+  /**
+   * Creates a WithoutReplacement distribution across the workers.
+   *
+   * @param tupleCounts
+   *          list of how many tuples each worker has.
+   * @param sampleSize
+   *          total number of samples to distribute across the workers.
+   * @return array representing the distribution across the workers.
+   */
+  private int[] withoutReplacementDistribution(List<Integer> tupleCounts,
+      int sampleSize) {
+    int[] distribution = new int[tupleCounts.size()];
+    int totalTupleCount = 0;
+    for (int val : tupleCounts) {
+      totalTupleCount += val;
+    }
+    List<Integer> logicalTupleCounts = new ArrayList<>(tupleCounts);
+
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount - i);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (int j = 0; j < logicalTupleCounts.size(); j++) {
+        if (sampleTupleIdx < logicalTupleCounts.get(j) + tupleOffset) {
+          distribution[j] += 1;
+          // Cannot sample the same tuple, so pretend it doesn't exist anymore.
+          logicalTupleCounts.set(j, logicalTupleCounts.get(j) - 1);
+          break;
+        }
+        tupleOffset += logicalTupleCounts.get(j);
+      }
+    }
+    return distribution;
+  }
+
+  /**
+   * Returns the sample size of this operator. If operator was created using a
+   * samplePercentage, this value will be 0 until after fetchNextReady.
+   */
+  public int getSampleSize() {
+    return sampleSize;
+  }
+
+  /** Returns whether this operator is using a percentage sample. */
+  public boolean isPercentageSample() {
+    return isPercentageSample;
+  }
+
+  /**
+   * Returns the percentage of total tuples that this operator will distribute.
+   * Will be 0 if the operator was created using a specific sampleSize.
+   */
+  public float getSamplePercentage() {
+    return samplePercentage;
+  }
+
+  /** Returns the type of sampling distribution that this operator will create. */
+  public SamplingType getSampleType() {
+    return sampleType;
+  }
+
+  @Override
+  public Schema generateSchema() {
+    return SCHEMA;
+  }
+
+  @Override
+  protected void init(final ImmutableMap<String, Object> execEnvVars) {
+    rand = new Random();
```

Comment:
As I said above, I think all of this can just be moved to the constructor, since nothing here is expensive and nothing can fail. `tupleCounts` and `streamCounts` could be initialized in their declarations.